### ISSUE TITLE
Locale loading inconsistency

### ISSUE THREAD

**Issue by h15ter** [moment/moment#3946]
moment().locale(['ru-custom', 'fr-custom', 'en', 'en-GB']).locale()
>> "fr"

moment().locale(['ru-custom', 'fr-custom', 'en-GB', 'en']).locale()
>> "en-gb"

Expected to be either first partial match, or first exact match.

**Comment by icambron** [comment]
I think it's refusing to consider `en-GB` in the first case because you already prefer the more specific `en`. Probably right for it to take `en-GB` in both case. Code is here: https://github.com/moment/moment/blob/develop/src/lib/locale/locales.js#L24

**Comment by MTelling** [comment]
So it should always choose the first correct locale it gets?

**Comment by MTelling** [comment]
In the tests https://github.com/moment/moment/blob/develop/src/test/moment/locale.js#L88
It says it should prefer earlier if that works. So should this be changed? 

I haven't contributed to this before, so hope it's okay I comment here.

**Comment by icambron** [comment]
Good catch; I hadn't remember that it was a conscious choice. But apparently it was. From the [pull request](https://github.com/moment/moment/pull/1142):

> On the next logic, if the code takes 'en-gb' over 'en' when passed ['en', 'en-gb'], that's a bug

Having started from scratch earlier today I came to the opposite conclusion for largely the same reason @h15ter did; it makes it a little bit harder to understand what's going on. But when I wrote this code in 2013, I spent a lot more time thinking about how this should work than I did this morning, and I went to some trouble to make this true. So the chances are that I was right then and wrong now. So I think I'll default to making this "won't fix" until someone either convinces me (or another maintainer) that it should change or I have a chance to really rethink it. 

Re: your edit: It considers 'fr' and 'en' to be equivalently specified by ['fr-custom' and 'en'], which seems right to me.

**Comment by MTelling** [comment]
If I may give my opinion, the way it currently works seems confusing. I can understand your reasoning about it but if i were to use the library, I would probably not understand the way it chooses. In my mind it would make sense to choose the first match in the list.

**Comment by h15ter** [comment]
@icambron It's not only about the tail of array actually. I probably messed things up a bit by listing all 4 locales, let me explain...

Here is a simplified example:
moment().locale(['ru-custom', 'fr-custom']).locale()
>>"fr"
moment().locale(['fr-custom', 'ru-custom']).locale()
>>"ru"
moment().locale(['fr-custom', 'ru-custom', 'es-custom']).locale()
>>"es"
So it always picks the last one, which is abviously wrong.

Source of problem is here:
https://github.com/moment/moment/blob/develop/src/lib/locale/locales.js#L37
The code automatically prefers the next locale (of the same 'split'-length) over current one irrespectively of being able to load it. It traverses the array in this way:
'fr-custom' -> try to load 'fr-custom.js' -> fail -> prefer next
'ru-custom' -> try to load 'ru-custom.js' -> fail -> prefer next
'es-custom' -> try to load 'es-custom.js' -> fail -> no next, the reduce 'j' counter
'es' -> try to load 'es.js' -> success!

Now it's clear why array tail matters (example from my initial post):
['fr-custom', 'en'] -> 'split'-length for just 'en' is smaller than for 'fr-custom', so the code reduces 'j' counter and loads 'fr' locale before moving over next one
['fr-custom', 'en-GB'] -> 'split'-length is the same, so code will move over 'en-GB' before even probing 'fr'

And finally:
moment().locale(['fr-FR', 'ru-RU', 'nb-NO', 'custom-custom']).locale()
>>"en"
We completely failed to load a locale, because we skipped all partial matches till the last one, and the last one can't be resolved at all.

**Comment by MTelling** [comment]
Wouldn't it make sense just to go over the locales twice? If there is a perfect match in the list then take that. Otherwise go through the list again, now only looking at the first part before the dash. Take the first match found. 
 
```
moment().locale(['ru-custom', 'fr-custom']).locale() -> 'ru'
moment().locale(['fr-custom', 'ru-custom']).locale() -> 'fr'
moment().locale(['fr-custom', 'ru-custom', 'es-custom']).locale() -> 'fr'
moment().locale(['ru-custom', 'fr-custom', 'en-GB', 'en']).locale() -> 'en-GB'
moment().locale(['ru-custom', 'fr-custom', 'en', 'en-GB']).locale() -> 'en'
```

This is something easily explained to users?

**Comment by h15ter** [comment]
@MTelling In my opinion first match should win, no matter if it's perfect or not. The whole idea behind passing locales as array is to try loading them in desired order and gracefully fallback if desired locale is not available. I can't even imagine situation when exact match matters more than order.

Common case: there is user's preferred locale (the one he set explicitly in his profile) and fallback locale (for instance, if 80% of your customers are Australian, you use 'en-AU' as a fallback). Order is everything here, no matter how perfect the match is. If user's preferred locale is 'ng-BO', you can't just fallback to 'en-AU' because it's a better match.

From library user's perspective, passing locales as an array is just a "syntax sugar":
```
let cultures = ['ng-NO', 'ru-RU', 'en-AU'], 
     locale = moment.localeData(cultures);
is expected to be an equivalent of
let cultures = ['ng-NO', 'ru-RU', 'en-AU'], 
     locale = cultures.map(c => moment.localeData(c)).find(locale => locale != null);
```

**Comment by MTelling** [comment]
@h15ter I see your point! I agree :)

**Comment by ovangle** [comment]
So I'm thinking the new chooseLocale algorithm should be:

1. Attempt to load all normalized strings in the order they're passed
2. Find all non-empty prefixes of the strings and attempt to load them in order
3. Repeat 2 until there are no non-empty prefixes

If that's the case, then it's a 5 minute change and non-breaking for 99% of cases.

**Comment by ovangle** [comment]
I added a quick fix for this and bundled it with the rest of the locale changes. I could easily do another fix for 2.x, but I've removed `chooseLocale` entirely as part of my other PR, so it'd be a pain to merge.

**Comment by marwahaha** [comment]
Up for grabs for a 2.x change.

**Comment by marwahaha** [comment]
Thanks for all the discussion.

This logic has been changed as of 2.25.3:
https://momentjs.com/docs/#/i18n/changing-locale/
```
Finally, Moment will search intelligently through an array of locales and their substrings.

moment.locale(['en-nz', 'en-au']); // 'en-au', not 'en'

The logic works as follows -- the next locale is picked and tried as-is. If that fails, the code normally tries to chop the last bit (normally the country designation) and try again. However, if the next array element has the same or longer prefix as the one to be tried, the iteration continues. So for example if the array has the sequence

"AA-BB", "AA-CC", "XX-YY"

then first "AA-BB" is tried, then a naive solution would try "AA", but this one instead checks to see that "AA-CC" is actually more concrete than "AA", so it tries "AA-CC" next, and only after it fails (if it fails) it tries "AA", because "XX-YY" does not have "AA" as prefix. So in the end the following locales are tried in this order (assuming all fail so the next one is tried):

"AA-BB", "AA-CC", "AA", "XX-YY", "XX"

```

Please re-open if you still have concerns.
